spring-cloud-stream-binders/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -542,62 +419,53 @@ public Object doWithRetry(RetryContext context) throws RuntimeException {
-		KafkaPropertiesAccessor accessor = new KafkaPropertiesAccessor(properties);
-
-		int maxConcurrency = accessor.getConcurrency(defaultConcurrency);
+	private Binding<MessageChannel> createKafkaConsumer(String name, final MessageChannel moduleInputChannel,
+														KafkaConsumerProperties properties, String group, long referencePoint) {
 
 		validateTopicName(name);
-
-		int numPartitions = accessor.getNumberOfKafkaPartitionsForConsumer();
-		Collection<Partition> allPartitions = ensureTopicCreated(name, numPartitions, defaultReplicationFactor);
+		int minKafkaPartitions = properties.getMinPartitionCount();
+		int instance = properties.getCount();
+		if (instance == 0) {
+			throw new IllegalArgumentException("Instance count cannot be zero");
+		}
+		int numPartitions = Math.max(minKafkaPartitions, instance * properties.getConcurrency());
+		Collection<Partition> allPartitions = ensureTopicCreated(name, numPartitions, replicationFactor);
 
 		Decoder<byte[]> valueDecoder = new DefaultDecoder(null);
 		Decoder<byte[]> keyDecoder = new DefaultDecoder(null);
 
 		Collection<Partition> listenedPartitions;
 
-		int moduleCount = accessor.getCount();
-
-		if (moduleCount == 1) {
+		if (instance == 1) {
 			listenedPartitions = allPartitions;
 		}
 		else {
-			listenedPartitions = new ArrayList<Partition>();
+			listenedPartitions = new ArrayList<>();
 			for (Partition partition : allPartitions) {
 				// divide partitions across modules
-				if (accessor.getPartitionIndex() != -1) {
-					if ((partition.getId() % moduleCount) == accessor.getPartitionIndex()) {
+				if (properties.isPartitioned()) {
+					if ((partition.getId() % instance) == properties.getPartitionIndex()) {
 						listenedPartitions.add(partition);
 					}
 				}
 				else {
-					int moduleSequence = accessor.getSequence();
-					if (moduleCount == 0) {
-						throw new IllegalArgumentException("The Kafka transport does not support 0-count modules");
-					}
-					else {
-						// sequence numbers are zero-based
-						if ((partition.getId() % moduleCount) == (moduleSequence - 1)) {
-							listenedPartitions.add(partition);
-						}
+					int instanceIndex = properties.getInstanceIndex();
+					// sequence numbers are zero-based
+					if ((partition.getId() % instance) == instanceIndex) {
+						listenedPartitions.add(partition);
 					}
 				}
 			}
 		}
 		topicsInUse.put(name, listenedPartitions);
-		ReceivingHandler rh = new ReceivingHandler();
+		ReceivingHandler rh = new ReceivingHandler(properties);
 		rh.setOutputChannel(moduleInputChannel);
 
 		final FixedSubscriberChannel bridge = new FixedSubscriberChannel(rh);
 		bridge.setBeanName("bridge." + name);
 
 		final KafkaMessageListenerContainer messageListenerContainer =
-				createMessageListenerContainer(accessor, group, maxConcurrency, listenedPartitions,
-						referencePoint);
+				createMessageListenerContainer(properties, group, null, listenedPartitions, referencePoint);
```

Comment:
Actually I see below... that is mutually exclusive with the `listenedPartitions`. It might be worth having 2 separate methods instead. WDYT?